### DISCUSSION TITLE
How to authenticate users after using `auth.login.before` hook?

### DISCUSSION THREAD

**Discussion by chiragsrvstv** [directus/directus#6368] [Archive]
Hello!
I'm using `auth.login.before` hook to do some stuff like verifying the bcrypt hash and updating the password but I'm not able to figure out how can I get out of the hook and proceed further to authenticate users. 

For instance, say I just wanna log the input in the hook and then authenticate the user, nothing really happens if I just return the input from the hook. And if I try using `authenticate()` from `AuthenticationService` then hook gets called again and the whole thing gets stuck into an indefinite loop.

https://user-images.githubusercontent.com/32494927/122545016-fd413b00-d04a-11eb-8454-202f180bd1db.mov

Here is the hook that I'm using:
```JS
module.exports = function registerHook({ services, exceptions }) {
  const { AuthenticationService } = services

  return {
    'auth.login.before': async function (input, options) {
      console.log('REACHED AUTH HOOK')
      console.log(input)
      
      // Do something
      return input
    },
  }
}
```

**Comment by rijkvanzanten** [comment]
> verifying the bcrypt hash and updating the password but I'm not able to figure out how can I get out of the hook and proceed further to authenticate users.

You might be able to use the UsersService to update the password hash, and then return the password hash from the authenticate hook. This'll make the authentication service use the updated password hash for the verification:

https://github.com/directus/directus/blob/3b7376020781ecefc825e935422286322f92e01a/api/src/services/authentication.ts#L125-L134

> And if I try using authenticate() from AuthenticationService then hook gets called again and the whole thing gets stuck into an indefinite loop.

This makes sense, as the hook is called in the authenticate method itself.

**Reply by chiragsrvstv** [reply]
Thanks @rijkvanzanten !
But even if I return the password hash from the authenticate hook, it still gives out a 401 http error with the message: `Invalid user credentials`.
Here is how I'm returning the password hash from the hook:
```js
// After verifying bcrypt hash successfully, update password
const user = new UsersService(options)
const update = await user.updateOne(options.user, {password: input.password})
const { password } = await user.knex('directus_users').where({email: input.email}).select('password').first()
return password
```
I know I could use UsersService to fetch the password hash but I still need to figure that out, so I've simply used knex.

**Reply by rijkvanzanten** [reply]
You have to return the same structure as the input, not just the password 🙂 

It's passing the original user object to the hook, and then replacing the user information with the result coming from the hook(s):

https://github.com/directus/directus/blob/bb86de93f05ae023d672cdef32316e8f413e1f49/api/src/services/authentication.ts#L62-L81

Then that output object is used for things like hash verification etc.

**Reply by chiragsrvstv** [reply]
Thanks @rijkvanzanten ! Returning the whole user object works. 
It'd be awesome if this information gets included in the docs. I'd love to work on it by opening up a PR in few days, I guess.

**Reply by rijkvanzanten** [reply]
> I'd love to work on it by opening up a PR in few days

Yes please!
